### ISSUE TITLE
Default analyses not executed for subscans

### ISSUE THREAD

**Issue by AUTProgram** [OxfordIonTrapGroup/ndscan#211]
When running a subscan through a higher-level fragment, the default analyses are not executed as expected. Consider the following code:

```python
from artiq.experiment import EnumerationValue

from ndscan.experiment import (Annotation, CustomAnalysis, ExpFragment,
    FloatChannel, FloatParam, LinearGenerator, make_fragment_scan_exp,
    setattr_subscan)


class BaseFragment(ExpFragment):
    def build_fragment(self):
        self.setattr_param(
            "param1",
            FloatParam,
            "Parameter 1",
            default=1.0)
        self.setattr_param(
            "param2",
            FloatParam,
            "Parameter 2",
            default=3.0)
        self.setattr_param(
            "param3",
            FloatParam,
            "Parameter 3",
            default=4.0)
        self.setattr_param(
            "param4",
            FloatParam,
            "Parameter 4",
            default=8.0)

        self.setattr_result(
            "res",
            FloatChannel)

    def run_once(self):
        self.res.push(self.param1.get()*10 + self.param2.get()*20
                      + self.param3.get()*60 + self.param4.get()*80)

    def get_default_analyses(self):
        fits = [CustomAnalysis([self.param1],
                               self._create_analyse_fn(self.param1),
                               [FloatChannel("x0_fit")]),
                CustomAnalysis([self.param2],
                               self._create_analyse_fn(self.param2),
                               [FloatChannel("x0_fit")]),
                CustomAnalysis([self.param3],
                               self._create_analyse_fn(self.param3),
                               [FloatChannel("x0_fit")]),
                CustomAnalysis([self.param4],
                               self._create_analyse_fn(self.param4),
                               [FloatChannel("x0_fit")])]
        return fits

    def _create_analyse_fn(self, scanned_param):
        def _analyse_scan(axis_values, result_values, analysis_results):
            x = axis_values[scanned_param]
            y = result_values[self.res]

            analysis_results["x0_fit"].push(-1.0)

            annotations = [
                Annotation("location",
                           coordinates={
                               scanned_param: analysis_results["x0_fit"]})]

            return annotations
        return _analyse_scan


ScanBaseFragment = make_fragment_scan_exp(BaseFragment)


class SubscanFragment(ExpFragment):
    def build_fragment(self):
        self.setattr_fragment("scanned_fragment", BaseFragment)

        self.arg = self.get_argument(
            "Argument of scan",
            EnumerationValue(["x", "y"]))

        setattr_subscan(
            self,
            "scan_params",
            self.scanned_fragment,
            [(self.scanned_fragment, "param1"),
             (self.scanned_fragment, "param2")])

    def run_once(self):
        if self.arg == "x":
            _, _, fit_results = self.scan_params.run(
                [(self.scanned_fragment.param1,
                  LinearGenerator(0, 100, 10, randomise_order=False))])
            print(fit_results)

        if self.arg == "y":
            _, _, fit_results = self.scan_params.run(
                [(self.scanned_fragment.param2,
                  LinearGenerator(0, 100, 10, randomise_order=False))])
            print(fit_results)


ScanSubscanFragment = make_fragment_scan_exp(SubscanFragment)
```

When trying to open the containing file via the explorer, it raises
```
root:Terminating with exception (AttributeError: 'NoneType' object has no attribute 'identity')
Traceback (most recent call last):
  File "/home/ion/artiq/artiq/master/worker_impl.py", line 398, in main
    examine(ExamineDeviceMgr, ExamineDatasetMgr, obj["file"])
  File "/home/ion/artiq/artiq/master/worker_impl.py", line 256, in examine
    cls = exp_class((device_mgr, dataset_mgr, argument_mgr, scheduler_defaults))
  File "/home/ion/artiq/artiq/language/environment.py", line 244, in __init__
    self.build(*args, **kwargs)
  File "/home/ion/artiq_pkgs/ndscan/ndscan/experiment/entry_point.py", line 482, in build
    max_transitory_error_retries=max_transitory_error_retries)
  File "/home/ion/artiq_pkgs/ndscan/ndscan/experiment/entry_point.py", line 79, in build
    self.fragment = fragment_init()
  File "/home/ion/artiq_pkgs/ndscan/ndscan/experiment/entry_point.py", line 480, in <lambda>
    super().build(lambda: fragment_class(self, [], *args),
  File "/home/ion/artiq/artiq/language/environment.py", line 244, in __init__
    self.build(*args, **kwargs)
  File "/home/ion/artiq_pkgs/ndscan/ndscan/experiment/fragment.py", line 75, in build
    self.build_fragment(*args, **kwargs)
  File "/home/ion/share/testing/bugs_mwe/ndscan_subscan_analysis.py", line 97, in build_fragment
    (self.scanned_fragment, "param2")])
  File "/home/ion/artiq_pkgs/ndscan/ndscan/experiment/subscan.py", line 264, in setattr_subscan
    analyses = filter_default_analyses(fragment, axes.values())
  File "/home/ion/artiq_pkgs/ndscan/ndscan/experiment/scan_runner.py", line 339, in filter_default_analyses
    if analysis.has_data(axis_identities):
  File "/home/ion/artiq_pkgs/ndscan/ndscan/experiment/default_analysis.py", line 199, in has_data
    for h in self._required_axis_handles) == set(scanned_axes)
  File "/home/ion/artiq_pkgs/ndscan/ndscan/experiment/default_analysis.py", line 199, in <genexpr>
    for h in self._required_axis_handles) == set(scanned_axes)
AttributeError: 'NoneType' object has no attribute 'identity'
```
This seems to happen since `setattr_subscan` filters all default analyses, but at this point only stores for the parameters passed to `setattr_subscan` have been created. So the remaining two CustomAnalyses fail when calling `has_data`.

When the latter two CustomAnalyses (for param3 and param4) have been commented out, the experiment opens in dashboard and runs without throwing an error. However, when choosing argument "x", the analysis for this parameter is not executed (fit_results returns an empty dictionary). This seems to be because the `required_axis` of that analysis is checked against both axes that could be scanned in the subscan (even when only one is actually scanned over) and therefore thinks it is not applicable to be used.

Trying to call `setattr_subscan` twice, one for `param1` and one for `param2`, cannot get around the problem as it fails, saying that the `res` result channel can only be used from one subscan.

Is this intended behaviour or a case that has not been considered before?

**Comment by dnadlinger** [comment]
Yes, that's definitely a bug. (Even if it was intended behaviour, it would need a better error message.)

Seems like the analyses should only be filtered in `setattr_subscan` if `expose_analysis_results` is `True`, and the filtering for `execute_default_analyses` postponed to runtime, where the actual axes are then known.

**Comment by AUTProgram** [comment]
Going through the code for `subscan.run`, I noticed the comment here
```
def _handle_default_analyses(self, axes, coordinate_sinks, always_run):
        if not self._analyses:
            return {}, {}

        # Re-filter analyses based on actual scan axes to support slightly dodgy use
        # case where a lower-dimensional scan is actually taken than originally
        # announced – should revisit this design.
        axis_identities = [(ax.param_schema["fqn"], ax.path) for ax in axes]
        analyses = [a for a in self._analyses if a.has_data(axis_identities)]
```
Why would this be considered a dodgy use case? In my situation, I would like to run two separate 1D subscans on different parameters for a subfragment, using the same result channel of the subfragment. So with the current implementation of `setattr_subscan`, I have to give it a list of the two parameters as `axis_params`, even though I am never intending to run an actual 2D scan. What's the recommended approach otherwise?

**Comment by AUTProgram** [comment]
> Just to make sure there are no misunderstandings, you can execute each subscan only once per `run_once`, otherwise the results have no place to go.

Yes, I'm setting the subscan for both parameters, but only actually run it for one of them. The choice is made based on an argument supplied by the user from the dashboard (basically an if-else with string comparison at the moment, which is not very elegant). Is there any workaround currently to run 1D subscans for both parameters in the same run_once of a top-level fragment?

**Comment by dnadlinger** [comment]
> Is there any workaround currently to run 1D subscans for both parameters in the same run_once of a top-level fragment?

Only manually creating all the subscan result channels/… twice, which is exceedingly annoying, or creating two of the fragment to be sub-scanned.
